### PR TITLE
Add converter from Soss's SampleChains

### LINKED ISSUE #124: Support recent Soss versions
Soss is currently at v0.18.3, but our docs currently only support v0.14. We should update to support v0.18 and ensure that the converter still works. This will involve adding support for Soss's [SampleChains.jl](https://github.com/cscherrer/SampleChains.jl). @cscherrer noted on Slack that some of the SampleChains functionality will be moved into [TupleVectors.jl](https://github.com/cscherrer/TupleVectors.jl), so it's probably best to wait until that has happened to work on this.

### PR DESCRIPTION
**Pull request by sethaxen**
This PR will add support for Soss's new sample wrapper, whose interface is defined in SampleChains.jl. It fixes #124. It works with very little additional code needed, as `AbstractChain` and `TupleVector` both look like `AbstractVector`s of `NamedTuple`s, for which we already have a converter.

~Right now it is missing a test suite and keywords for providing other groups~. So far already the example in Soss ReadMe works:

```julia
julia> using ArviZ, Soss, MeasureTheory, SampleChainsDynamicHMC

julia> m = @model x begin
           α ~ Lebesgue(ℝ)
           β ~ Normal()
           σ ~ Exponential()
           y ~ For(x) do xj
               Normal(α + β * xj, σ)
           end
           return y
       end;

julia> x = randn(20);

julia> predα = predictive(m, :α);

julia> y = rand(predα(x=x,α=10.0));

julia> post = sample(DynamicHMCChain, m(x=x) | (y=y,))
4000-element MultiChain with 4 chains and schema (σ = Float64, β = Float64, α = Float64)
(σ = 0.0928±0.016, β = -1.053±0.028, α = 10.031±0.018)

julia> idata = convert_to_inference_data(post)
InferenceData with groups:
	> posterior
	> sample_stats

julia> summarystats(idata)
3×10 DataFrame
 Row │ variable  mean     sd       hdi_3%   hdi_97%  mcse_mean  mcse_sd  ess_bulk  ess_tail  r_hat   
     │ String    Float64  Float64  Float64  Float64  Float64    Float64  Float64   Float64   Float64 
─────┼───────────────────────────────────────────────────────────────────────────────────────────────
   1 │ α          10.031    0.018    9.996   10.064      0.0      0.0      3859.0    3629.0     1.01
   2 │ σ           0.093    0.016    0.067    0.125      0.002    0.001     544.0     125.0     1.05
   3 │ β          -1.053    0.028   -1.105   -1.002      0.002    0.001     234.0      73.0     1.05

julia> idata.sample_stats
Dataset (xarray.Dataset)
Dimensions:          (chain: 4, draw: 1000)
Coordinates:
  * chain            (chain) int64 0 1 2 3
  * draw             (draw) int64 0 1 2 3 4 5 6 ... 993 994 995 996 997 998 999
Data variables:
    acceptance_rate  (chain, draw) float64 0.5217 0.8707 0.743 ... 0.9272 0.8404
    n_steps          (chain, draw) int64 3 3 3 3 3 3 3 7 3 ... 7 3 3 3 7 7 3 3 3
    diverging        (chain, draw) bool False False False ... False False False
    lp               (chain, draw) float64 32.86 35.41 34.35 ... 35.17 33.72
    tree_depth       (chain, draw) int64 2 2 2 2 2 2 2 3 2 ... 2 2 2 2 3 3 1 2 2
    turning          (chain, draw) bool True True True True ... True True True
Attributes:
    created_at:     2021-05-18T21:23:24.910798
    arviz_version:  0.11.2
```

cc @cscherrer

### CODE REVIEW

**Comment by sethaxen** [comment]
Alright, this PR is feature-complete. I'll handle nested posterior draws more generally in a future PR.

@cscherrer, could you review this and see if anything looks off to you? In particular, can the Soss example in the quickstart be cleaned up any more using functions in the Soss API? And is there a way to evaluate the log-likelihood elementwise on `y` instead of for all of `y`? (see e.g. the Turing and CmdStan examples in the quickstart). Also note the wild rank plots, which indicate the issue we discussed on Slack.

**Comment by cscherrer** [comment]
Thanks @sethaxen ! This is great, but I think we should hold off a bit until the next SampleChains release, since I'm refactoring things to take a config with keyword arguments. That PR is https://github.com/cscherrer/SampleChainsDynamicHMC.jl/pull/6, currently failing because it depends on the master branch of SampleChains (I'm makeing a release for that now).

**Comment by sethaxen** [comment]
Great! If you can, just let me know when the release happens so I can update this PR.

File: src/samplechains.jl [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+end
+
+@require SampleChainsDynamicHMC = "6d9fd711-e8b2-4778-9c70-c1dfb499d4c4" begin
+    using .SampleChainsDynamicHMC: SampleChainsDynamicHMC
+
+    function _samplechains_info(chain::SampleChainsDynamicHMC.DynamicHMCChain)
+        info = SampleChains.info(chain)
+        termination = info.termination
+        tree_stats = (
+            lp=info.π,
```
Comment (sethaxen) [inline src/samplechains.jl:44]: On further inspection, `π` is [documented](https://github.com/tpapp/DynamicHMC.jl/blob/v3.1.0/src/NUTS.jl#L207-L208) as the log density of the proposed phasepoint, i.e. the energy. It's not clear if the value of the log density for that sample is returned anywhere.

```suggestion
            energy=info.π,
```

**Comment by sethaxen** [comment]
> Thanks @sethaxen ! This is great, but I think we should hold off a bit until the next SampleChains release, since I'm refactoring things to take a config with keyword arguments. That PR is [cscherrer/SampleChainsDynamicHMC.jl#6](https://github.com/cscherrer/SampleChainsDynamicHMC.jl/pull/6), currently failing because it depends on the master branch of SampleChains (I'm makeing a release for that now).

@cscherrer is everything pretty stable with SampleChainsDynamicHMC.jl now? I'd like to finish up this PR.

**Comment by cscherrer** [comment]
I think so, and @kskyten has asked about using ArviZ from Soss for his JuliaCon workshop, so the timing is perfect. Thank you!

**Comment by sethaxen** [comment]
Great! I'll wrap up in the next few days. It'd be great if you could review it when it's ready, but I know you have a lot on your plate right now, so no worries if not.

**Comment by ahartikainen** [comment]
Is there a test for multidimensional objects?

**Comment by sethaxen** [comment]
> Is there a test for multidimensional objects?

No, good point, we need one.

**Comment by sethaxen** [comment]
@cscherrer this is ready for review now. In particular, if you could check that the [quickstart example](https://arviz-devs.github.io/ArviZ.jl/previews/PR131/quickstart/#Plotting-with-Soss.jl-outputs) is using the simplest expressions available in Soss, that would be great.

Also, does SampleChainsDynamicHMC store the log density of all draws? I could only find the log density of the last draw.

**Comment by cscherrer** [comment]
Thanks Seth!

Trying this out, it confused me for a minute that some variables aren't defined. It would help to have each section self-contained, even if that makes the document a little longer.

So far I think you need to add `using Random` and
```julia
J = 8
y = [28.0, 8.0, -3.0, 7.0, -1.0, 1.0, 18.0, 12.0]
σ = [15.0, 10.0, 16.0, 11.0, 9.0, 11.0, 10.0, 18.0]
schools = [
    "Choate",
    "Deerfield",
    "Phillips Andover",
    "Phillips Exeter",
    "Hotchkiss",
    "Lawrenceville",
    "St. Paul's",
    "Mt. Hermon",
];

nwarmup, nsamples, nchains = 1000, 1000, 4;
```

Log-density should be there, it's a bug if not. I'll look in a minutes

**Comment by cscherrer** [comment]
Not really a Soss thing, but to me tau (especially a positive one) usually hints that it will be used as a precision. Maybe this is always how 8 schools is written? I forget. If that's the case then no worries :)

**Comment by cscherrer** [comment]
The `For` expression
```
y ~ For(1:J) do j
        Normal(μ = θ[j], σ = σ[j])
    end
```
can also be written
```
y ~ For(θ,σ) do θj, σj
        Normal(θj, σj)
    end
```

The arguments are `zip`ped together and then traversed. Doesn't matter much, just another option

**Comment by cscherrer** [comment]
Looks like I don't have a nice representation for a vector of TupleVectors, I'll need to update that.

It's maybe a matter of personal preference, but I think it's a little easier to understand this:
```julia
prior_priorpred = [rand(rng, param_mod, nsamples) for _ in 1:nchains]
```

Most of these are all really minor things, maybe more style differences than anything. Generally I think it's looking good, and the plots look great! Guess I'll need to start using Arviz :)

**Comment by cscherrer** [comment]
Log density is here, but we need a better way to get to it:
```julia
help?> SampleChainsDynamicHMC.DynamicHMC.TreeStatisticsNUTS
  struct TreeStatisticsNUTS

  Diagnostic information for a single tree built with the No-U-turn sampler.

  Fields
  ≡≡≡≡≡≡≡≡

  Accessing fields directly is part of the API.

    •  π
       Log density (negative energy).

    •  depth
       Depth of the tree.

    •  termination
       Reason for termination. See InvalidTree and REACHED_MAX_DEPTH.

    •  acceptance_rate
       Acceptance rate statistic.

    •  steps
       Number of leapfrog steps evaluated.

    •  directions
       Directions for tree doubling (useful for debugging).

julia> SampleChainsDynamicHMC.info(getchains(post)[1])[1]
DynamicHMC.TreeStatisticsNUTS(-29.960754670347434, 6, turning at positions -108:-115, 0.9942311322062368, 119, DynamicHMC.Directions(0xa9581304))
```

**Comment by cscherrer** [comment]
Oh sorry I missed this one. `predict` is there for convenience, basically calling `predictive` on the right things, followed by `rand`. So you can do
```julia
postpred = [predict(mod, c) for c in getchains(post)]
```

I need to make this easier, `predict(mod, multichain)` should just produce another multichain. Or something like that.

**Comment by sethaxen** [comment]
Thanks for looking at this!

> Trying this out, it confused me for a minute that some variables aren't defined. It would help to have each section self-contained, even if that makes the document a little longer.

I may change this eventually, but it's outside the scope of this PR.

> Not really a Soss thing, but to me tau (especially a positive one) usually hints that it will be used as a precision. Maybe this is always how 8 schools is written? I forget. If that's the case then no worries :)

Yeah, that's right, for 8 schools, it's usually written `tau`.

> Log density is here, but we need a better way to get to it:

I'm actually okay with where that is, but that's the log density of the _Hamiltonian_, i.e. negative of the sum of potential and kinetic energy. This is a little vaguely documented here, but that's what it works out to be. I'm looking for the (unnormalized) log density of the posterior, that is, negative of the potential energy.

> `predict` is there for convenience, basically calling `predictive` on the right things, followed by `rand`. So you can do
> 
> ```julia
> postpred = [predict(mod, c) for c in getchains(post)]
> ```

Ah, good to know. But it looks like `predict` doesn't allow providing the `rng` argument, so I'll use `rand` for now.

> I need to make this easier, `predict(mod, multichain)` should just produce another multichain. Or something like that.

Yes, that makes sense to me.

> Guess I'll need to start using Arviz :)

😄

**Pull request merged**